File: pset1/credit.py
def is_valid_input(input):
    try:
        int(input)
        return True
    except:
        return False
        
def get_company(cardNumber):
    if cardNumber.startswith('4'):
        return 'VISA'
    elif int(cardNumber[:2]) in range(50,56):
        return 'MASTERCARD'
    elif cardNumber.startswith('31') or cardNumber.startswith('37'):
        return 'AMEX'
        
def sum_doubled_digit(digit):
    doubled = digit * 2
    if doubled < 10:
        return doubled
    else:
        return 1 + (doubled % 10)
        
def toggle_bool(value):
    if value == True:
        return False
    else:
        return True
        
def is_valid_number(card_number):
    sum_doubled_digits = 0
    sum_other_digits = 0
    reversed_card_number = card_number[::-1]
    to_be_doubled = False
    for digit in reversed_card_number:
        if to_be_doubled:
            sum_doubled_digits += sum_doubled_digit(int(digit))
        else:
            sum_other_digits += int(digit)
        to_be_doubled = toggle_bool(to_be_doubled)
    if (sum_doubled_digits + sum_other_digits) % 10 == 0:
        return True
    else:
        return False
    
if __name__ == "__main__":
    card_number = input("Number: ")
    while not is_valid_input(card_number):
        card_number = input("Number: ")
        
    if is_valid_number(card_number):
        print(get_company(card_number))
    else:
        print("INVALID, but company would be {}".format(get_company(card_number)))

File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>

bool isValidInput(string input);
string getCompany(string cardNumber);
int sumDoubledDigit(int digit);
bool toggleBool(bool value);
bool isValidNumber(string cardNumber);
string getFirstDigits(string inputString, int numDigits);
int getLength(string inputString);
string reverseString(string inputString);
int charToInt(char character);

int main(void)
{
    printf("Number: ");
    string cardNumber = get_string();
    while (!isValidInput(cardNumber)) {
        printf("Retry: ");
        cardNumber = get_string();
    }
    
    if (isValidNumber(cardNumber)) {
        printf("%s\n", getCompany(cardNumber));
    } else {
        printf("INVALID\n");
    }
}

int getFirstTwoDigits(string inputString) {
    int firstTwo = (10 * charToInt(inputString[0])) + charToInt(inputString[1]);
    return firstTwo;
}

string getCompany(string cardNumber) {
    int firstTwo = getFirstTwoDigits(cardNumber);
    if (firstTwo >= 40 && firstTwo < 50) {
        return "VISA";
    } else if (firstTwo > 50 && firstTwo <= 55) {
        return "MASTERCARD";
    } else if (firstTwo == 31 || firstTwo == 37) {
        return "AMEX";
    } else {
        return "INVALID";
    }
}

int sumDoubledDigit(int digit) {
    int doubled = digit * 2;
    if (doubled < 10) {
        return doubled;
    } else {
        return 1 + (doubled % 10);
    }
}

bool toggleBool(bool value) {
    if (value == true) {
        return false;
    } else {
        return true;
    }
}

bool isValidNumber(string cardNumber) {
    int length = getLength(cardNumber);
    int sumDoubledDigits = 0;
    int sumOtherDigits = 0;
    //Reverse order of Card Number so digits can be easily alternated
    char reversedCardNumber[length];
    int reversedIndex = length - 1;
    for (int i = 0; i < length; i++) {
        reversedCardNumber[i] = cardNumber[reversedIndex];
        reversedIndex -= 1;
    }
    //Calculate checksum
    bool toBeDoubled = false;
    for (int i = 0; i < length; i++) {
        if (toBeDoubled) {
            sumDoubledDigits += sumDoubledDigit(charToInt(reversedCardNumber[i]));
        } else {
            sumOtherDigits += charToInt(reversedCardNumber[i]);
        }
        toBeDoubled = toggleBool(toBeDoubled);
    }
    if ((sumDoubledDigits + sumOtherDigits) % 10 == 0) {
        return true;
    } else {
        return false;
    }
}

int getLength(string inputString) {
    int length = 0;
    while (inputString[length]) {
        length += 1;
    }
    return length;
}

int charToInt(char character) {
    return character - '0';
}

bool isValidInput(string input) {
    int length = getLength(input);
    int digitCount = 0;
    for (int i = 0; i < length; i++) {
        if (input[i] >= 48 && input[i] <= 57) {
            digitCount += 1;
        } else {
            return false;
        }
    }
    if (digitCount > 0) {
        return true;
    } else {
        return false;
    }
}